File: backend/rest-blog/src/main/java/io/github/mateuszuran/restblog/service/TagsService.java
package io.github.mateuszuran.restblog.service;

import io.github.mateuszuran.restblog.exception.PostNotFoundException;
import io.github.mateuszuran.restblog.exception.TagNotFoundException;
import io.github.mateuszuran.restblog.model.Post;
import io.github.mateuszuran.restblog.model.Tags;
import io.github.mateuszuran.restblog.repository.PostRepository;
import io.github.mateuszuran.restblog.repository.TagsRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Stream;

@Slf4j
@Service
public class TagsService {
    private final TagsRepository repository;
    private final PostRepository postRepository;

    public TagsService(final TagsRepository repository, final PostRepository postRepository) {
        this.repository = repository;
        this.postRepository = postRepository;
    }

    public Tags addTagToPost(Long id, Tags tag) {
        Tags newTag = new Tags();
        newTag.toUpdate(tag);
        newTag.setPost(postRepository.findById(id).orElseThrow(() -> new PostNotFoundException(id)));
        return repository.save(newTag);
    }

    public Tags getTag(Long postId, Long tagId) {
        return findTagInPost(postId, tagId);
    }

    public Tags updateTagContent(Long postId, Long tagId, Tags tagToUpdate) {
        Tags tagFromDb = findTagInPost(postId, tagId);
        tagFromDb.setContent(tagToUpdate.getContent());
        return repository.save(tagFromDb);
    }

    public void deleteTag(Long postId, Long tagId) {
        var result = findTagInPost(postId, tagId);
        repository.deleteById(result.getId());
    }

    private Tags findTagInPost(Long postId, Long tagId) {
        return repository.findAllByPostId(postId)
                .stream().filter(tag ->
                        tag.getId().equals(tagId)).findFirst()
                .orElseThrow(() -> new TagNotFoundException(tagId));
    }
}

File: frontend/src/services/tags.service.js
import api from './api'

const addTag = (id, tags) => {
  return api.post("/add-tag", tags, { params: { id: id } });
};

const getTag = (id, tagId) => {
  return api.get("/tag", { params: { id: id, tagId: tagId } });
};

const editTag = (id, tagId, tags) => {
  return api.put("/edit-tag", tags, { params: { id: id, tagId: tagId } });
};

const deleteTag = (id, tagId) => {
  return api.delete("/delete-tag", { params: { id: id, tagId: tagId } });
};

const TagsService = {
  addTag,
  getTag,
  editTag,
  deleteTag
};
export default TagsService;